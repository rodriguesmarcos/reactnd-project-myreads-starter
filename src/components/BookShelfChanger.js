import React from 'react';
import PropTypes from 'prop-types';

const BookShelfChanger = ({book, shelves, onChangeShelf, history}) => (
  <select
    defaultValue={book.shelf || 'none'}
    onChange={(e) => {
      const shelf = e.target.value;
      onChangeShelf(book, shelf);
      if ( history )
        history.push("/");      
    }}
  >
    <option value="none" disabled>Move to...</option>
    {shelves.map(shelf => (
      <option
        key={shelf.id}
        value={shelf.id}
      >{shelf.name}</option>
    ))}
    <option value="none">None</option>
  </select>
);

BookShelfChanger.propTypes = {
  shelves: PropTypes.array.isRequired,
  book: PropTypes.shape({
    id: PropTypes.string.isRequired
  }).isRequired,
  onChangeShelf: PropTypes.func.isRequired,
}

export default BookShelfChanger;
