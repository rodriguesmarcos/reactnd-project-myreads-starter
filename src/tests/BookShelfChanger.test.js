import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import BookShelfChanger from '../components/BookShelfChanger.js';
import books from './fixtures/books';
import shelves from './fixtures/shelves';

describe('<BookShelfChanger />', () => {

  const minProps = {
    book: books[0],
    shelves: shelves,
    onChangeShelf: () => {},
  }

  test('renders without exploding', () => {
    const wrapper = shallow(<BookShelfChanger {...minProps}/>);
    expect(wrapper.length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should call onChangeShelf', () => {
    const onChangeShelfSpy = jest.fn();
    const value = shelves[2].id;
    const wrapper = shallow(<BookShelfChanger {...minProps} onChangeShelf={onChangeShelfSpy}/>);
    const select = wrapper.find('select');
    select.simulate('change', {
      target: { value }
    });
    expect(onChangeShelfSpy).toHaveBeenLastCalledWith(books[0],value);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
