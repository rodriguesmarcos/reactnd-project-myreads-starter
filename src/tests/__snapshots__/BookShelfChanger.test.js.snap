exports[`<BookShelfChanger /> renders without exploding 1`] = `
<select
  defaultValue="currentlyReading"
  onChange={[Function]}>
  <option
    disabled={true}
    value="none">
    Move to...
  </option>
  <option
    key="currentlyReading"
    value="currentlyReading">
    Currently Reading
  </option>
  <option
    key="wantToRead"
    value="wantToRead">
    Want to Read
  </option>
  <option
    key="read"
    value="read">
    Read
  </option>
  <option
    value="none">
    None
  </option>
</select>
`;

exports[`<BookShelfChanger /> should call onChangeShelf 1`] = `
<select
  defaultValue="currentlyReading"
  onChange={[Function]}>
  <option
    disabled={true}
    value="none">
    Move to...
  </option>
  <option
    key="currentlyReading"
    value="currentlyReading">
    Currently Reading
  </option>
  <option
    key="wantToRead"
    value="wantToRead">
    Want to Read
  </option>
  <option
    key="read"
    value="read">
    Read
  </option>
  <option
    value="none">
    None
  </option>
</select>
`;

exports[`<BookShelfChanger /> should render Loading spinner 1`] = `
<select
  defaultValue="currentlyReading"
  onChange={[Function]}>
  <option
    disabled={true}
    value="none">
    Move to...
  </option>
  <option
    key="currentlyReading"
    value="currentlyReading">
    Currently Reading
  </option>
  <option
    key="wantToRead"
    value="wantToRead">
    Want to Read
  </option>
  <option
    key="read"
    value="read">
    Read
  </option>
  <option
    value="none">
    None
  </option>
</select>
`;
